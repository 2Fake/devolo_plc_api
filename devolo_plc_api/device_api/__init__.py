"""The devolo device API."""
from .deviceapi import DeviceApi

SERVICE_TYPE = "_dvl-deviceapi._tcp.local."

__all__ = ["DeviceApi"]
