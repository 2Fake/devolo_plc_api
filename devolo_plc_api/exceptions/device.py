"""Exceptions that can occur when communicating with a devolo device."""


class DeviceNotFound(Exception):
    """The device was not found."""

    def __init__(self, ip: str) -> None:
        """Initialize error."""
        super().__init__(f"The device {ip} did not answer.")


class DeviceUnavailable(Exception):
    """The device is not available, e.g. in standby."""

    def __init__(self) -> None:
        """Initialize error."""
        super().__init__("The device is currently not available. Maybe on standby?")


class DevicePasswordProtected(Exception):
    """The device is password protected."""

    def __init__(self) -> None:
        """Initialize error."""
        super().__init__("The used password is wrong.")
