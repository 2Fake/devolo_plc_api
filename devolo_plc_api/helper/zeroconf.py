from zeroconf import ServiceBrowser, ServiceStateChange, Zeroconf
import time
import socket


def _on_service_state_change(zeroconf: Zeroconf, service_type: str, name: str, state_change: ServiceStateChange):
    """ Service handler for Zeroconf state changes. """
    if state_change is ServiceStateChange.Added:
        zeroconf.get_service_info(service_type, name)


def get_token(ip, service_name):
    # TODO: Optimize the for & if
    zeroconf = Zeroconf()
    browser = ServiceBrowser(zeroconf, service_name, [_on_service_state_change])
    start_time = time.time()
    while not time.time() > start_time + 10:
        for mdns_name in zeroconf.cache.entries():
            try:
                if hasattr(mdns_name, "server"):
                    zc = zeroconf.cache.cache.get(mdns_name.key)
                    for item in zc:
                        if hasattr(item, "server"):
                            try:
                                zc2 = zeroconf.cache.cache[item.server.lower()]
                                for item2 in zc2:
                                    try:
                                        if socket.inet_ntoa(item2.address) == ip:
                                            zc3 = zeroconf.cache.cache.get(item.key)
                                            for item3 in zc3:
                                                if hasattr(item3, "text"):
                                                    parsed_text = parse_zeroconf_text(item3.text)
                                                    if service_name == '_dvl-plcnetapi._tcp.local.':
                                                        return parsed_text['Path']
                                                    else:
                                                        return parsed_text['Path'], parsed_text.get('Features').split(",")
                                    except OSError:
                                        continue
                            except KeyError:
                                continue
            except (AttributeError, OSError, ValueError):
                continue
    browser.cancel()
    zeroconf.close()


def parse_zeroconf_text(text):
    entries = {}
    total_length = len(text)
    parsed_length = 0
    while parsed_length < total_length:
        entry_length = int(text[parsed_length])
        entry = text[parsed_length + 1:parsed_length + entry_length + 1].decode('UTF-8')
        parsed_length = parsed_length + entry_length + 1
        split_entry = entry.split('=')
        entries[split_entry[0]] = split_entry[1]
    entries["Path"] = entries["Path"].split("/")[0]
    return entries
