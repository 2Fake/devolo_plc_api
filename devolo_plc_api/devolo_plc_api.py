from aiohttp import ClientSession

from .device_api.device_api import DeviceApi
from .plc_net_api.plc_net_api import PlcNetApi


class DevoloPlcApi():
    def __init__(self, ip: str, session: ClientSession):
        self.device_api = DeviceApi(ip, session)
        self.plc_net_api = PlcNetApi(ip, session)
