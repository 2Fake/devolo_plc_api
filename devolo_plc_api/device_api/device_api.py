from .wifi_network.wifi_network import WifiNetwork
from ..helper.zeroconf import get_token


class DeviceApi:
    def __init__(self, ip, session):
        self.token, self.features = get_token(ip, "_dvl-deviceapi._tcp.local.")
        feature_mapping = {"wifi1": WifiNetwork}

        for feature in self.features:
            try:
                setattr(self, feature, feature_mapping.get(feature)(ip, session, 14791, self.token))
            except TypeError:
                continue