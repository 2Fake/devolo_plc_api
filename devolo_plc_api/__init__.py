"""The devolo PLC API."""
from importlib.metadata import PackageNotFoundError, version

from .device import Device

try:
    __version__ = version("devolo_plc_api")
except PackageNotFoundError:
    # package is not installed - e.g. pulled and run locally
    __version__ = "0.0.0"

__all__ = ["Device", "__version__"]
