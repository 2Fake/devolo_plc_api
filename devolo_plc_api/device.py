"""Representation of your devolo device."""
from __future__ import annotations

import asyncio
import ipaddress
import logging
import struct
from contextlib import suppress
from datetime import date
from types import TracebackType

import httpx
from zeroconf import DNSQuestionType, ServiceInfo, ServiceStateChange, Zeroconf
from zeroconf.asyncio import AsyncServiceBrowser, AsyncServiceInfo, AsyncZeroconf

from .device_api import SERVICE_TYPE as DEVICEAPI, DeviceApi
from .exceptions.device import DeviceNotFound
from .plcnet_api import DEVICES_WITHOUT_PLCNET, SERVICE_TYPE as PLCNETAPI, PlcNetApi
from .zeroconf import ZeroconfServiceInfo


class Device:  # pylint: disable=too-many-instance-attributes
    """
    Representing object for your devolo PLC device. It stores all properties and functionalities discovered during setup.

    :param ip: IP address of the device to communicate with.
    :param plcnetapi: Reuse externally gathered data for the plcnet API
    :param deviceapi: Reuse externally gathered data for the device API
    :param zeroconf_instance: Zeroconf instance to be potentially reused.
    """

    MDNS_TIMEOUT = 300

    def __init__(
        self,
        ip: str,
        zeroconf_instance: AsyncZeroconf | Zeroconf | None = None,
    ) -> None:
        """Initialize the device."""
        self.ip = ip
        self.mac = ""
        self.mt_number = "0"
        self.product = ""
        self.technology = ""
        self.serial_number = "0"

        self.device: DeviceApi | None = None
        self.plcnet: PlcNetApi | None = None

        self._background_tasks: set[asyncio.Task] = set()
        self._browser: dict[str, AsyncServiceBrowser] = {}
        self._connected = False
        self._info: dict[str, ZeroconfServiceInfo] = {PLCNETAPI: ZeroconfServiceInfo(), DEVICEAPI: ZeroconfServiceInfo()}
        self._logger = logging.getLogger(f"{self.__class__.__module__}.{self.__class__.__name__}")
        self._multicast = False
        self._password = ""
        self._session_instance: httpx.AsyncClient | None = None
        self._zeroconf_instance = zeroconf_instance
        logging.captureWarnings(True)

        self._session: httpx.AsyncClient
        self._zeroconf: AsyncZeroconf

    def __del__(self) -> None:
        """Warn user, if the connection was not properly closed."""
        if self._connected and self._session_instance is None:
            self._logger.warning("Please disconnect properly from the device.")

    async def __aenter__(self) -> Device:
        """Connect to a device asynchronously when entering a context manager."""
        await self.async_connect()
        return self

    async def __aexit__(self, exc_type: type | None, exc_val: BaseException | None, exc_tb: TracebackType | None) -> None:
        """Disconnect to a device asynchronously when entering a context manager."""
        await self.async_disconnect()

    def __enter__(self) -> Device:
        """Connect to a device synchronously when leaving a context manager."""
        self.connect()
        return self

    def __exit__(self, exc_type: type | None, exc_val: BaseException | None, exc_tb: TracebackType | None) -> None:
        """Disconnect to a device synchronously when leaving a context manager."""
        self.disconnect()

    @property
    def firmware_date(self) -> date:
        """Date the firmware was built."""
        return date.fromisoformat(self._info[DEVICEAPI].properties.get("FirmwareDate", "1970-01-01")[:10])

    @property
    def firmware_version(self) -> str:
        """Firmware version currently installed."""
        return self._info[DEVICEAPI].properties.get("FirmwareVersion", "")

    @property
    def hostname(self) -> str:
        """mDNS hostname of the device."""
        return self._info[DEVICEAPI].hostname

    @property
    def password(self) -> str:
        """The currently set device password."""
        return self._password

    @password.setter
    def password(self, password: str) -> None:
        """Change the currently set device password."""
        self._password = password
        if self.device:
            self.device.password = password
        if self.plcnet:
            self.plcnet.password = password

    async def async_connect(self, session_instance: httpx.AsyncClient | None = None) -> None:
        """
        Connect to a device asynchronous.

        :param: session_instance: Session client instance to be potentially reused.
        """
        self._session_instance = session_instance
        self._session = self._session_instance or httpx.AsyncClient()
        if not self._zeroconf_instance:
            self._zeroconf = AsyncZeroconf()
        elif isinstance(self._zeroconf_instance, Zeroconf):
            self._zeroconf = AsyncZeroconf(zc=self._zeroconf_instance)
        else:
            self._zeroconf = self._zeroconf_instance
        await asyncio.gather(self._get_device_info(), self._get_plcnet_info())
        if not self.device and not self.plcnet:
            raise DeviceNotFound(self.ip)
        self._connected = True

    def connect(self) -> None:
        """Connect to a device synchronous."""
        asyncio.run(self.async_connect())

    async def async_disconnect(self) -> None:
        """Disconnect from a device asynchronous."""
        if self._connected:
            for browser in self._browser.values():
                await browser.async_cancel()
            if not self._zeroconf_instance:
                await self._zeroconf.async_close()
            if not self._session_instance:
                await self._session.aclose()
            self._connected = False

    def disconnect(self) -> None:
        """Disconnect from a device synchronous."""
        asyncio.run(self.async_disconnect())

    async def _get_device_info(self) -> None:
        """Get information from the devolo Device API."""
        service_type = DEVICEAPI
        await self._get_zeroconf_info(service_type=service_type)
        if not self._info[service_type].properties:
            await self._retry_zeroconf_info(service_type=service_type)
        if self._info[service_type].properties:
            self.mt_number = self._info[service_type].properties.get("MT", "0")
            self.product = self._info[service_type].properties.get("Product", "")
            self.serial_number = self._info[service_type].properties["SN"]
            self.device = DeviceApi(ip=self.ip, session=self._session, info=self._info[service_type])
            self.device.password = self.password

    async def _get_plcnet_info(self) -> None:
        """Get information from the devolo PlcNet API."""
        service_type = PLCNETAPI
        if self.mt_number in DEVICES_WITHOUT_PLCNET:
            return
        await self._get_zeroconf_info(service_type=service_type)
        if not self._info[service_type].properties and self.mt_number not in DEVICES_WITHOUT_PLCNET:
            await self._retry_zeroconf_info(service_type=service_type)
        if self._info[service_type].properties:
            self.mac = self._info[service_type].properties["PlcMacAddress"]
            self.technology = self._info[service_type].properties.get("PlcTechnology", "")
            self.plcnet = PlcNetApi(ip=self.ip, session=self._session, info=self._info[service_type])
            self.plcnet.password = self.password

    async def _get_zeroconf_info(self, service_type: str) -> None:
        """Browse for the desired mDNS service types and query them."""
        self._logger.debug("Browsing for %s", service_type)
        counter = 0
        addr = None if self._multicast else self.ip
        question_type = DNSQuestionType.QM if self._multicast else DNSQuestionType.QU
        self._browser[service_type] = AsyncServiceBrowser(
            zeroconf=self._zeroconf.zeroconf,
            type_=service_type,
            handlers=[self._state_change],
            addr=addr,
            question_type=question_type,
        )
        while not self._info[service_type].properties and counter < self.MDNS_TIMEOUT:
            counter += 1
            await asyncio.sleep(0.01)

    async def _retry_zeroconf_info(self, service_type: str) -> None:
        """Retry getting the zeroconf info using multicast."""
        self._logger.debug(
            "Having trouble getting %s via unicast messages. Switching to multicast for this device.", service_type
        )
        self._multicast = True
        await self._get_zeroconf_info(service_type=service_type)

    def _state_change(self, zeroconf: Zeroconf, service_type: str, name: str, state_change: ServiceStateChange) -> None:
        """Evaluate the query result."""
        if state_change == ServiceStateChange.Removed:
            return
        task = asyncio.create_task(self._get_service_info(zeroconf, service_type, name))
        self._background_tasks.add(task)
        task.add_done_callback(self._background_tasks.remove)

    async def _get_service_info(self, zeroconf: Zeroconf, service_type: str, name: str) -> None:
        """Get service information, if IP matches."""
        service_info = AsyncServiceInfo(service_type, name)
        question_type = DNSQuestionType.QM if self._multicast else DNSQuestionType.QU
        with suppress(RuntimeError):
            await service_info.async_request(zeroconf, timeout=1000, question_type=question_type)

        if not service_info.addresses or str(ipaddress.ip_address(service_info.addresses[0])) != self.ip:
            return  # No need to continue, if there are no relevant service information

        self._logger.debug("Updating service info of %s for %s", service_type, service_info.server_key)
        if info := self.info_from_service(service_info):
            self._info[service_type] = info

    @staticmethod
    def info_from_service(service_info: ServiceInfo) -> ZeroconfServiceInfo | None:
        """Return prepared info from mDNS entries."""
        properties = {}
        if not service_info.addresses:
            return None  # No need to continue, if there is no IP address to contact the device

        total_length = len(service_info.text)
        offset = 0
        while offset < total_length:
            (parsed_length,) = struct.unpack_from("!B", service_info.text, offset)
            key_value = service_info.text[offset + 1 : offset + 1 + parsed_length].decode("UTF-8").split("=")
            properties[key_value[0]] = key_value[1]
            offset += parsed_length + 1

        return ZeroconfServiceInfo(
            address=service_info.addresses[0],
            hostname=service_info.server,
            port=service_info.port,
            properties=properties,
        )
