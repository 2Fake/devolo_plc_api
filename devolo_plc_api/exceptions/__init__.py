"""Exceptions used by the package."""
from .device import DeviceNotFound, DevicePasswordProtected, DeviceUnavailable
from .feature import FeatureNotSupported

__all__ = ["DeviceNotFound", "DevicePasswordProtected", "DeviceUnavailable", "FeatureNotSupported"]
