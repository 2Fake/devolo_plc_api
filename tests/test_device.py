"""Test communicating with a devolo device."""
from datetime import date
from unittest.mock import AsyncMock, Mock, patch

import pytest
from zeroconf import ServiceStateChange

from devolo_plc_api.device import EMPTY_INFO, Device
from devolo_plc_api.device_api import SERVICE_TYPE as DEVICEAPI
from devolo_plc_api.device_api import DeviceApi
from devolo_plc_api.exceptions.device import DeviceNotFound
from devolo_plc_api.plcnet_api import DEVICES_WITHOUT_PLCNET
from devolo_plc_api.plcnet_api import SERVICE_TYPE as PLCNETAPI
from devolo_plc_api.plcnet_api import PlcNetApi
from tests.mocks.mock_zeroconf import MockServiceBrowser

from . import TestData
from .mocks.mock_device import state_change
from .stubs.zeroconf import StubAsyncServiceInfo


@pytest.mark.usefixtures("block_communication")
class TestDevice:
    """Test devolo_plc_api.device.Device class."""

    @pytest.mark.parametrize("feature", [""])
    def test_set_password(self, mock_device: Device, device_api: DeviceApi):
        """Test setting a device password is also reflected in the device API."""
        mock_device.device = device_api
        mock_device.password = "super_secret"
        assert mock_device.device.password == "super_secret"

    @pytest.mark.asyncio
    @pytest.mark.parametrize("feature", [""])
    async def test_async_connect(self, mock_device: Device, device_api: DeviceApi, plcnet_api: PlcNetApi):
        """Test that connecting to a device calls methos to collect information from the APIs."""
        with patch("devolo_plc_api.device.Device._get_device_info") as gdi, patch(
            "devolo_plc_api.device.Device._get_plcnet_info"
        ) as gpi:
            mock_device.device = device_api
            mock_device.plcnet = plcnet_api
            await mock_device.async_connect()
            assert gdi.call_count == 1
            assert gpi.call_count == 1
            assert getattr(mock_device, "_connected")

            await mock_device.async_connect(session_instance=AsyncMock())
            assert gdi.call_count == 2
            assert gpi.call_count == 2
            assert getattr(mock_device, "_connected")

    @pytest.mark.asyncio
    async def test_async_connect_not_found(self, mock_device: Device):
        """Test that an exception is raised if both APIs are not available."""
        with patch("devolo_plc_api.device.Device._get_device_info"), patch(
            "devolo_plc_api.device.Device._get_plcnet_info"
        ), pytest.raises(DeviceNotFound):
            await mock_device.async_connect()
            assert not getattr(mock_device, "_connected")

    def test_connect(self, mock_device: Device):
        """Test that the sync connect method just calls the async connect method."""
        with patch("devolo_plc_api.device.Device.async_connect", new=AsyncMock()) as ac:
            mock_device.connect()
            assert ac.call_count == 1

    @pytest.mark.asyncio
    # pylint: disable=protected-access
    async def test_async_disconnect(self, mock_device: Device):
        """Test that disconnecting from a device cleans up Zeroconf and the HTTP client."""
        await mock_device.async_connect()
        await mock_device.async_disconnect()
        assert mock_device._zeroconf.async_close.call_count == 1  # type: ignore
        assert mock_device._session.aclose.call_count == 1  # type: ignore
        assert not getattr(mock_device, "_connected")

        await mock_device.async_connect(session_instance=AsyncMock())
        await mock_device.async_disconnect()
        assert mock_device._zeroconf.async_close.call_count == 1  # type: ignore
        assert mock_device._session.aclose.call_count == 0  # type: ignore
        assert not getattr(mock_device, "_connected")

    def test_disconnect(self, mock_device: Device):
        """Test that the sync disconnect method just calls the async disconnect method."""
        with patch("devolo_plc_api.device.Device.async_disconnect", new=AsyncMock()) as ad:
            mock_device.disconnect()
            assert ad.call_count == 1

    @pytest.mark.asyncio
    @pytest.mark.usefixtures("mock_service_browser")
    async def test_async_context_manager(self, test_data: TestData):
        """Test the async context manager."""
        with patch("devolo_plc_api.device.Device._state_change", state_change):
            async with Device(test_data.ip) as device:
                assert getattr(device, "_connected")
            assert not getattr(device, "_connected")

    @pytest.mark.usefixtures("mock_service_browser")
    def test_context_manager(self, test_data: TestData):
        """Test the sync context manager."""
        with patch("devolo_plc_api.device.Device._state_change", state_change):
            with Device(test_data.ip) as device:
                assert getattr(device, "_connected")
            assert not getattr(device, "_connected")

    @pytest.mark.asyncio
    @pytest.mark.usefixtures("mock_device_api")
    async def test__get_device_info(self, mock_device: Device, test_data: TestData):
        """Test that information from the device API are filled in."""
        with patch("devolo_plc_api.device.Device._get_zeroconf_info"):
            device_info = test_data.device_info[DEVICEAPI]
            mock_device.password = "super_secret"
            await mock_device.async_connect()
            assert mock_device.firmware_date == date.fromisoformat(device_info["properties"]["FirmwareDate"])
            assert mock_device.firmware_version == device_info["properties"]["FirmwareVersion"]
            assert mock_device.serial_number == device_info["properties"]["SN"]
            assert mock_device.mt_number == device_info["properties"]["MT"]
            assert mock_device.product == device_info["properties"]["Product"]
            assert isinstance(mock_device.device, DeviceApi)
            assert mock_device.device.password == mock_device.password

    @pytest.mark.asyncio
    async def test__get_device_info_multicast(self, test_data: TestData):
        """Test that devices having trouble with unicast zeroconf are queried twice."""
        with patch("devolo_plc_api.device.Device._get_zeroconf_info") as gzi, pytest.raises(DeviceNotFound):
            device = Device(test_data.ip)
            await device.async_connect()
            assert getattr(device, "_multicast")
            assert gzi.call_count == 2

    @pytest.mark.asyncio
    @pytest.mark.usefixtures("mock_plcnet_api")
    async def test__get_plcnet_info(self, mock_device: Device, test_data: TestData):
        """Test that information from the plcnet API are filled in."""
        with patch("devolo_plc_api.device.Device._get_zeroconf_info"):
            device_info = test_data.device_info[PLCNETAPI]
            await mock_device.async_connect()
            assert mock_device.mac == device_info["properties"]["PlcMacAddress"]
            assert mock_device.technology == device_info["properties"]["PlcTechnology"]
            assert isinstance(mock_device.plcnet, PlcNetApi)

    @pytest.mark.asyncio
    @pytest.mark.parametrize("feature", [""])
    async def test__not_get_plcnet_info(self, mock_device: Device, device_api: DeviceApi):
        """Test that devices know to not have a plcnet API don't query it."""
        mock_device.mt_number = DEVICES_WITHOUT_PLCNET[0]
        with patch("devolo_plc_api.device.Device._get_zeroconf_info") as gzi, patch(
            "devolo_plc_api.device.Device._get_device_info"
        ):
            mock_device.device = device_api
            await mock_device.async_connect()
            assert gzi.call_count == 0
            assert not mock_device.plcnet

    @pytest.mark.asyncio
    async def test__get_plcnet_info_multicast(self, test_data: TestData):
        """Test that devices having trouble with unicast zeroconf are queried twice."""
        with patch("devolo_plc_api.device.Device._get_zeroconf_info") as gzi, pytest.raises(DeviceNotFound):
            device = Device(test_data.ip)
            await device.async_connect()
            assert getattr(device, "_multicast")
            assert gzi.call_count == 2

    @pytest.mark.asyncio
    # pylint: disable=protected-access
    async def test__get_zeroconf_info_timeout(self, mock_device: Device):
        """Test that the mDNS service browser times out after 3 seconds."""
        mock_device._info["_http._tcp.local."] = EMPTY_INFO
        await mock_device.async_connect()
        with patch("devolo_plc_api.device.AsyncServiceBrowser._async_start"), patch(
            "devolo_plc_api.device.AsyncServiceBrowser._async_cancel"
        ), patch("asyncio.sleep") as sleep:
            await mock_device._get_zeroconf_info("_http._tcp.local.")
            assert sleep.call_count == 300

    @pytest.mark.asyncio
    async def test__get_zeroconf_info_device_info_exists(self, mock_device: Device, mock_service_browser: MockServiceBrowser):
        """Test early return if information already exist."""
        await mock_device.async_connect()
        assert mock_service_browser.async_cancel.call_count == 0

    @pytest.mark.asyncio
    @pytest.mark.usefixtures("mock_service_browser")
    async def test__state_change_no_service_info(self, test_data: TestData):
        """Test that waiting for mDNS responses continues, if no service info were received."""
        with patch("devolo_plc_api.device.Device.info_from_service") as ifs, patch(
            "devolo_plc_api.device.AsyncServiceInfo.async_request"
        ), patch("asyncio.sleep"):
            mock_device = Device(ip=test_data.ip, plcnetapi=None, deviceapi=test_data.device_info[DEVICEAPI])
            await mock_device.async_connect()
            assert ifs.call_count == 0

    @pytest.mark.asyncio
    # pylint: disable=protected-access
    async def test__state_change_removed(self, mock_device: Device):
        """Test that service information are not processed on state change to removed."""
        with patch("devolo_plc_api.device.Device._retry_zeroconf_info"), patch(
            "devolo_plc_api.device.Device._get_service_info"
        ) as gsi:
            mock_device._state_change(Mock(), PLCNETAPI, PLCNETAPI, ServiceStateChange.Removed)
            assert gsi.call_count == 0

    @pytest.mark.asyncio
    @pytest.mark.usefixtures("mock_service_browser")
    async def test__get_service_info(self, test_data: TestData):
        """Test storing of information discovered via mDNS."""
        with patch("devolo_plc_api.device.AsyncServiceInfo", StubAsyncServiceInfo), patch(
            "devolo_plc_api.device.PlcNetApi"
        ), patch("devolo_plc_api.device.AsyncServiceInfo.async_request") as ar:
            mock_device = Device(ip=test_data.ip, plcnetapi=None, deviceapi=test_data.device_info[DEVICEAPI])
            await mock_device.async_connect()
            assert ar.call_count == 2
            assert mock_device.mac == test_data.device_info[PLCNETAPI]["properties"]["PlcMacAddress"]

    def test_info_from_service_no_address(self, mock_device: Device):
        """Test ignoring information received for an other address."""
        service_info = Mock()
        service_info.addresses = None
        assert mock_device.info_from_service(service_info) == {}
