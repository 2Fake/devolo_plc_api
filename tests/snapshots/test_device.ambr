# serializer version: 1
# name: TestDevice.test_async_connect
  Device(
    MDNS_TIMEOUT=300,
    device=DeviceApi(
      features=list([
        'wifi1',
      ]),
      password='',
      url='http://192.0.2.1:None/1234567890abcdef/v0/',
    ),
    firmware_date=datetime.date(2020, 6, 29),
    firmware_version='5.5.1',
    hostname='device.local',
    ip='192.0.2.1',
    mac='AABBCCDDEEFF',
    mt_number='2730',
    password='',
    plcnet=PlcNetApi(
      password='',
      url='http://192.0.2.1:None/1234567890abcdef/v0/',
    ),
    product='dLAN pro 1200+ WiFi ac',
    serial_number='1234567890123456',
    technology='hpav',
  )
# ---
# name: TestDevice.test_async_connect.1
  Device(
    MDNS_TIMEOUT=300,
    device=DeviceApi(
      features=list([
        'wifi1',
      ]),
      password='',
      url='http://192.0.2.1:None/1234567890abcdef/v0/',
    ),
    firmware_date=datetime.date(2020, 6, 29),
    firmware_version='5.5.1',
    hostname='device.local',
    ip='192.0.2.1',
    mac='AABBCCDDEEFF',
    mt_number='2730',
    password='',
    plcnet=PlcNetApi(
      password='',
      url='http://192.0.2.1:None/1234567890abcdef/v0/',
    ),
    product='dLAN pro 1200+ WiFi ac',
    serial_number='1234567890123456',
    technology='hpav',
  )
# ---
# name: TestDevice.test_async_connect_plc[DeviceType.PLC]
  Device(
    MDNS_TIMEOUT=300,
    device=DeviceApi(
      features=list([
        'wifi1',
      ]),
      password='',
      url='http://192.0.2.1:80/1234567890abcdef/v0/',
    ),
    firmware_date=datetime.date(2020, 6, 29),
    firmware_version='5.5.1',
    hostname='device.local',
    ip='192.0.2.1',
    mac='AABBCCDDEEFF',
    mt_number='3046',
    password='',
    plcnet=PlcNetApi(
      password='',
      url='http://192.0.2.1:80/1234567890abcdef/v0/',
    ),
    product='dLAN pro 1200+ WiFi ac',
    serial_number='1234567890123456',
    technology='hpav',
  )
# ---
# name: TestDevice.test_async_connect_plc[DeviceType.PLC].1
  Device(
    MDNS_TIMEOUT=300,
    device=DeviceApi(
      features=list([
        'wifi1',
      ]),
      password='',
      url='http://192.0.2.1:80/1234567890abcdef/v0/',
    ),
    firmware_date=datetime.date(2020, 6, 29),
    firmware_version='5.5.1',
    hostname='device.local',
    ip='192.0.2.1',
    mac='AABBCCDDEEFF',
    mt_number='3046',
    password='',
    plcnet=PlcNetApi(
      password='',
      url='http://192.0.2.1:80/1234567890abcdef/v0/',
    ),
    product='dLAN pro 1200+ WiFi ac',
    serial_number='1234567890123456',
    technology='hpav',
  )
# ---
# name: TestDevice.test_async_connect_repeater[DeviceType.REPEATER]
  Device(
    MDNS_TIMEOUT=300,
    device=DeviceApi(
      features=list([
        'wifi1',
      ]),
      password='',
      url='http://192.0.2.1:80/1234567890abcdef/v0/',
    ),
    firmware_date=datetime.date(2020, 6, 29),
    firmware_version='5.5.1',
    hostname='device.local',
    ip='192.0.2.1',
    mac='',
    mt_number='3046',
    password='',
    plcnet=None,
    product='dLAN pro 1200+ WiFi ac',
    serial_number='1234567890123456',
    technology='',
  )
# ---
